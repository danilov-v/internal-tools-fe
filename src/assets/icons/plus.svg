<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg xmlns="http://www.w3.org/2000/svg" width="17.998" height="17.998" viewBox="0 0 4.762 4.762">
  <circle cx="2.381" cy="2.381" r="2.235" opacity=".94" fill="none" stroke="#f2f2f2" stroke-width=".292"/>
  <path d="M2.381.91v2.941M.911 2.377l2.94.007" opacity=".94" fill="none" stroke="#f2f2f2" stroke-width=".237"/>
</svg>
