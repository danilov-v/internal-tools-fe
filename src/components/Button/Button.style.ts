import styled from 'styled-components';

type DropDownButtonProps = {
  expanded?: boolean;
};

export const Button = styled.button`
  border: 1px solid #b3b3b3;
  border-radius: 2px;
  display: inline-block;
  background-color: transparent;
  color: #b3b3b3;
  cursor: pointer;
`;

export const DropDownButton = styled(Button)<DropDownButtonProps>`
  position: relative;
  padding-right: 15px;

  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 5px;
    right: 2px;
    width: 8px;
    height: 7px;
    background-color: #b3b3b3;
    clip-path: polygon(50% 100%, 0 0, 100% 0);
    transform: ${({ expanded }) =>
      expanded ? 'rotate(180deg)' : 'rotate(0deg)'};
  }
`;
