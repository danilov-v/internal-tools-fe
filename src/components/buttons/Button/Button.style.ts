import styled from 'styled-components';

type ContainerProps = {
  color: 'default' | 'primary' | 'secondary' | 'yellow';
  disabled: boolean;
  fullWidth: boolean;
  variant: 'contained' | 'outlined' | 'text';
};

const VARIANTS = {
  contained: {
    background: {
      default: 'primaryGray',
      primary: 'secondaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    border: 'solid 1px',
    color: {
      default: 'primaryWhite',
      primary: 'primaryWhite',
      secondary: 'primaryWhite',
      yellow: 'primaryWhite',
    },
    fontWeight: 400,
  },
  outlined: {
    background: {
      default: 'transparent',
      primary: 'transparent',
      secondary: 'transparent',
      yellow: 'transparent',
    },
    border: 'solid 1px',
    color: {
      default: 'primaryGray',
      primary: 'secondaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    fontWeight: 300,
  },
  text: {
    background: {
      default: 'transparent',
      primary: 'transparent',
      secondary: 'transparent',
      yellow: 'transparent',
    },
    border: 'none',
    color: {
      default: 'primaryGray',
      primary: 'primaryGreen',
      secondary: 'primaryRed',
      yellow: 'primaryYellow',
    },
    fontWeight: 400,
  },
};

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'background',
):
  | 'primaryGray'
  | 'primaryGreen'
  | 'primaryRed'
  | 'transparent'
  | 'primaryYellow';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'color',
):
  | 'primaryGray'
  | 'primaryGreen'
  | 'primaryRed'
  | 'primaryWhite'
  | 'primaryYellow';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'border',
): 'none' | 'solid 1px';

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'fontWeight',
): number;

function getVariantProperty(
  variant: 'contained' | 'outlined' | 'text',
  color: 'default' | 'primary' | 'secondary' | 'yellow',
  property: 'background' | 'border' | 'color' | 'fontWeight',
): string | number {
  return property === 'border' || property === 'fontWeight'
    ? VARIANTS[variant][property]
    : VARIANTS[variant][property][color];
}

export const Container = styled.button<ContainerProps>`
  align-items: center;
  background-color: ${({ color, theme, variant }) =>
    theme[getVariantProperty(variant, color, 'background')]};
  border-radius: 4px;
  border: ${({ color, theme, variant }) =>
    getVariantProperty(variant, color, 'border') === 'none'
      ? getVariantProperty(variant, color, 'border')
      : `${getVariantProperty(variant, color, 'border')} ${
          theme[getVariantProperty(variant, color, 'color')]
        }`};
  color: ${({ color, theme, variant }) =>
    theme[getVariantProperty(variant, color, 'color')]};
  cursor: ${({ disabled }) => (disabled ? 'default' : 'pointer')};
  display: flex;
  font-size: 18px;
  font-weight: ${({ color, variant }) =>
    getVariantProperty(variant, color, 'fontWeight')};
  justify-content: center;
  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  padding: 6px 12px;
  width: ${({ fullWidth }) => (fullWidth ? '100%;' : 'auto')};

  &:hover {
    opacity: ${({ disabled }) => (disabled ? 0.6 : 0.8)};
  }
`;

type TextProps = {
  endIcon: boolean;
  startIcon: boolean;
};

export const Text = styled.span<TextProps>`
  margin-left: ${({ startIcon }) => (startIcon ? 8 : 0)}px;
  margin-right: ${({ endIcon }) => (endIcon ? 8 : 0)}px;
`;
