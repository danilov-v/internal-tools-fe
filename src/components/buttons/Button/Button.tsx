import React, { MouseEvent, ReactNode } from 'react';

import * as S from './Button.style';

export const Button: React.FC<ButtonProps> = ({
  children = '',
  className = '',
  color = 'default',
  disabled = false,
  endIcon = null,
  fullWidth = false,
  name = '',
  onClick = () => {},
  startIcon = null,
  type = 'button',
  variant = 'outlined',
}) => (
  <S.Container
    className={className}
    color={color}
    disabled={disabled}
    fullWidth={fullWidth}
    name={name}
    onClick={onClick}
    type={type}
    variant={variant}
  >
    {startIcon}
    <S.Text startIcon={!!startIcon} endIcon={!!endIcon}>
      {children}
    </S.Text>
    {endIcon}
  </S.Container>
);

export type ButtonProps = {
  className?: string;
  color?: 'default' | 'primary' | 'secondary' | 'yellow';
  disabled?: boolean;
  endIcon?: ReactNode;
  fullWidth?: boolean;
  name?: string;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  startIcon?: ReactNode;
  type?: 'button' | 'submit' | 'reset';
  variant?: 'contained' | 'outlined' | 'text';
  children?: React.ReactNode;
};
