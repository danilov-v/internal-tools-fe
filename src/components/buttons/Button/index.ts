import { Button } from './Button';
import { withDropDownIndicator } from './HOC/withDropDownIndicator';

const DropDownButton = withDropDownIndicator(Button);

export { Button, DropDownButton };
