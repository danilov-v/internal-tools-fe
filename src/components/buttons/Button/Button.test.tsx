import React from 'react';
import { render, RenderResult } from '@testing-library/react';

import 'jest-styled-components';

import { ReactComponent as Icon } from 'assets/icons/plus.svg';
import { Button } from './Button';

const getComponent = (props = {}): RenderResult => {
  const parsedProps = {
    ...props,
  };

  return render(<Button {...parsedProps} />);
};

describe('Button component', () => {
  it('matches snapshot with only required props', () => {
    expect(getComponent().baseElement).toMatchSnapshot();
  });

  it('matches snapshots with additional props', () => {
    const containedPrimaryButton = getComponent({
      children: 'children',
      color: 'primary',
      disabled: true,
      type: 'button',
      variant: 'contained',
    });
    const outlinedSecondaryButton = getComponent({
      color: 'secondary',
      fullWidth: true,
      name: 'name',
      type: 'submit',
      variant: 'outlined',
    });
    const textYellowButton = getComponent({
      children: 'children',
      color: 'yellow',
      startIcon: <Icon />,
      type: 'reset',
      variant: 'text',
    });

    expect(containedPrimaryButton.baseElement).toMatchSnapshot();
    expect(outlinedSecondaryButton.baseElement).toMatchSnapshot();
    expect(textYellowButton.baseElement).toMatchSnapshot();
  });
});
