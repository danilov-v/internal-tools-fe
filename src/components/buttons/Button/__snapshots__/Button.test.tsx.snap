// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot with only required props 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  border: solid 1px undefined;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 1;
  padding: 6px 12px;
  width: auto;
}

.c0:hover {
  opacity: 0.8;
}

.c1 {
  margin-left: 0px;
  margin-right: 0px;
}

<body>
  <div>
    <button
      class="c0"
      color="default"
      name=""
      type="button"
    >
      <span
        class="c1"
      />
    </button>
  </div>
</body>
`;

exports[`Button component matches snapshots with additional props 1`] = `
<body>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  border: solid 1px undefined;
  cursor: default;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  font-weight: 400;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0.6;
  padding: 6px 12px;
  width: auto;
}

.c0:hover {
  opacity: 0.6;
}

.c1 {
  margin-left: 0px;
  margin-right: 0px;
}

<div>
    <button
      class="c0"
      color="primary"
      disabled=""
      name=""
      type="button"
    >
      <span
        class="c1"
      >
        children
      </span>
    </button>
  </div>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  border: solid 1px undefined;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  font-weight: 300;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 1;
  padding: 6px 12px;
  width: 100%;
}

.c0:hover {
  opacity: 0.8;
}

.c1 {
  margin-left: 0px;
  margin-right: 0px;
}

<div>
    <button
      class="c0"
      color="secondary"
      name="name"
      type="submit"
    >
      <span
        class="c1"
      />
    </button>
  </div>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  font-weight: 400;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 1;
  padding: 6px 12px;
  width: auto;
}

.c0:hover {
  opacity: 0.8;
}

.c1 {
  margin-left: 8px;
  margin-right: 0px;
}

<div>
    <button
      class="c0"
      color="yellow"
      name=""
      type="reset"
    >
      <svg>
        plus.svg
      </svg>
      <span
        class="c1"
      >
        children
      </span>
    </button>
  </div>
</body>
`;

exports[`Button component matches snapshots with additional props 2`] = `
<body>
  <div>
    <button
      class="sc-AxjAm hJZLSv"
      color="primary"
      disabled=""
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ dOafwD"
      >
        children
      </span>
    </button>
  </div>
  <div>
    <button
      class="sc-AxjAm eRuwuL"
      color="secondary"
      name="name"
      type="submit"
    >
      <span
        class="sc-AxirZ dOafwD"
      />
    </button>
  </div>
  <div>
    <button
      class="sc-AxjAm iNKxSg"
      color="yellow"
      name=""
      type="reset"
    >
      <svg>
        plus.svg
      </svg>
      <span
        class="sc-AxirZ gBmPvz"
      >
        children
      </span>
    </button>
  </div>
</body>
`;

exports[`Button component matches snapshots with additional props 3`] = `
<body>
  <div>
    <button
      class="sc-AxjAm hJZLSv"
      color="primary"
      disabled=""
      name=""
      type="button"
    >
      <span
        class="sc-AxirZ dOafwD"
      >
        children
      </span>
    </button>
  </div>
  <div>
    <button
      class="sc-AxjAm eRuwuL"
      color="secondary"
      name="name"
      type="submit"
    >
      <span
        class="sc-AxirZ dOafwD"
      />
    </button>
  </div>
  <div>
    <button
      class="sc-AxjAm iNKxSg"
      color="yellow"
      name=""
      type="reset"
    >
      <svg>
        plus.svg
      </svg>
      <span
        class="sc-AxirZ gBmPvz"
      >
        children
      </span>
    </button>
  </div>
</body>
`;
